lib/vrl/stdlib/src/hmac.rs

```diff
@@ -0,0 +1,226 @@
+pub struct Hmac;
+
+impl Function for Hmac {
+    fn identifier(&self) -> &'static str {
+        "hmac"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+
+            Parameter {
+                keyword: "key",
+                kind: kind::BYTES,
+                required: true,
+            },
+
+            Parameter {
+                keyword: "algorithm",
+                kind: kind::BYTES,
+                required: false
+            },
+
+            Parameter {
+                keyword: "encoding",
+                kind: kind::BYTES,
+                required: false
+            }
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "default (SHA-256, base64-encoded result)",
+                source: r#"hmac("Hello there", "supersecretkey")"#,
+                result: Ok("kmpc79vrb6SODvg4LwivUnb443+IhR9SSW55KcBPKo8="),
+            },
+
+            Example {
+                title: "SHA-256, hex encoded result",
+                source: r#"hmac("Hello there", "supersecretkey", encoding: "hex")"#,
+                result: Ok("926a5cefdbeb6fa48e0ef8382f08af5276f8e37f88851f52496e7929c04f2a8f")
+            },
+
+            Example {
+                title: "SHA1, base64-encoded result",
+                source: r#"hmac("Hello there", "supersecretkey", algorithm: "SHA1")"#,
+                result: Ok("795HKoopDtOb45EOUroxeHz1OWo=")
+            }
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let key = arguments.required("key");
+        let algorithm = arguments
+            .optional_enum("algorithm", &algorithms())?
+            .unwrap_or_else(|| value!("SHA-256"))
+            .try_bytes()
+            .expect("algorithm not bytes");
+        let encoding = arguments
+            .optional_enum("encoding", &encode_formats())?
+            .unwrap_or_else(|| value!("base64"))
+            .try_bytes()
+            .expect("encoding not bytes");
+
+        Ok(HmacFn { value, key, algorithm, encoding }.as_expr())
+    }
+}
+
+#[derive(Debug, Clone)]
+struct HmacFn {
+    value: Box<dyn Expression>,
+    key: Box<dyn Expression>,
+    algorithm: Bytes,
+    encoding: Bytes,
+}
+
+impl FunctionExpression for HmacFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let key = self.key.resolve(ctx)?;
+        let algorithm = &self.algorithm;
+        let encoding = &self.encoding;
+
+        hmac(value, key, algorithm, encoding)
+    }
+
+    fn type_def(&self, _: &state::TypeState) -> TypeDef {
+        TypeDef::bytes().infallible()

```

Comment:
This needs to be `fallible` since the algorithm or encoding chosen could be invalid. However, for the "happy path" where the algorithm and encoding are literals / known at compile-time, this can still be `infallible`.

`Expression::as_value` will try to evaluate an expression at compile-time, and you can check if the value returned from that is valid. (`HmacFn` contains the key and value expression so they can be accessed from the `type_def` function).